### ISSUE TITLE
Leverage setHours to have a more efficient implementation of start-end-of

### ISSUE THREAD

**Issue by aymericbouzy** [moment/moment#3619]
**Description of the Issue and Steps to Reproduce:**

```
let d = new Date()
d.setHours(0, 0, 0, 0)
```
is way more efficient than : 
```
let d = new Date()
d.setHours(0)
d.setMinutes(0)
d.setSeconds(0)
d.setMilliseconds(0)
```
(about **400** times faster)

When calling `startOf(unit)`, where `unit` is one of `'hour'`, `'day'`, `'week'`, `'month'`, `'year'`, moment should use the first method rather than the second one.

**Environment:**

I'm running a ReactNative app on an iOS Simulator, running iOS 10.1.

**Other information that may be helpful:**
* The time zone setting of the machine the code is running on : UTC + 01:00 (Paris, France)
* The time and date at which the code was run : 2016-11-23T10:56:19+01:00
* Other libraries in use (TypeScript, Immutable.js, etc) : ReactNative

```js
console.log( (new Date()).toString())
// [Log] Wed Nov 23 2016 10:58:47 GMT+0100 (CET) (index.ios.bundle, line 788)
console.log((new Date()).toLocaleString())
// [Log] 11/23/2016, 10:58:47 AM (index.ios.bundle, line 788)
console.log( (new Date()).getTimezoneOffset())
// [Log] -60 (index.ios.bundle, line 788)
console.log( navigator.userAgent)
// [Log] undefined (index.ios.bundle, line 788)
console.log(moment.version)
// [Log] 2.14.1 (index.ios.bundle, line 788)
```

**Comment by aymericbouzy** [comment]
#3620

**Comment by maggiepint** [comment]
closed in favor of #3620
